File: rust/src/numbers.rs
use crate::types::{Index, Size};

pub fn get_uint32(bytes: &[u8], offset: usize) -> u32 {
    u32::from_be_bytes([
        bytes[offset],
        bytes[offset + 1],
        bytes[offset + 2],
        bytes[offset + 3],
    ])
}

pub fn set_uint32(bytes: &mut [u8], offset: usize, number: u32) {
    bytes[offset..offset + 4].copy_from_slice(number.to_be_bytes().as_slice());
}

pub fn get_uint64(bytes: &[u8], offset: usize) -> u64 {
    u64::from_be_bytes([
        bytes[offset],
        bytes[offset + 1],
        bytes[offset + 2],
        bytes[offset + 3],
        bytes[offset + 4],
        bytes[offset + 5],
        bytes[offset + 6],
        bytes[offset + 7],
    ])
}

pub fn set_uint64(bytes: &mut [u8], offset: usize, number: u64) {
    bytes[offset..offset + 8].copy_from_slice(number.to_be_bytes().as_slice());
}

pub fn read_index(index: Index) -> u32 {
    get_uint32(&index, 0)
}

pub fn read_size(size: Size) -> usize {
    get_uint64(&size, 0) as usize
}

pub fn set_size(bytes: &mut Size, size: usize) {
    set_uint64(bytes, 0, size as u64)
}

File: cplusplus/src/numbers.c
#include "numbers.h"

#include "autograph.h"

uint32_t get_uint32(const uint8_t *bytes, const size_t offset) {
  uint32_t number = ((uint32_t)bytes[offset] << 24) |
                    ((uint32_t)bytes[offset + 1] << 16) |
                    ((uint32_t)bytes[offset + 2] << 8) | bytes[offset + 3];
  return number;
}

void set_uint32(uint8_t *bytes, const size_t offset, const uint32_t number) {
  bytes[offset] = (number >> 24) & 0xFF;
  bytes[offset + 1] = (number >> 16) & 0xFF;
  bytes[offset + 2] = (number >> 8) & 0xFF;
  bytes[offset + 3] = number & 0xFF;
}

uint64_t get_uint64(const uint8_t *bytes, const size_t offset) {
  uint64_t number =
      ((uint64_t)bytes[offset] << 56) | ((uint64_t)bytes[offset + 1] << 48) |
      ((uint64_t)bytes[offset + 2] << 40) |
      ((uint64_t)bytes[offset + 3] << 32) |
      ((uint64_t)bytes[offset + 4] << 24) |
      ((uint64_t)bytes[offset + 5] << 16) | ((uint64_t)bytes[offset + 6] << 8) |
      (uint64_t)bytes[offset + 7];
  return number;
}

void set_uint64(uint8_t *bytes, const size_t offset, const uint64_t number) {
  bytes[offset] = (number >> 56) & 0xFF;
  bytes[offset + 1] = (number >> 48) & 0xFF;
  bytes[offset + 2] = (number >> 40) & 0xFF;
  bytes[offset + 3] = (number >> 32) & 0xFF;
  bytes[offset + 4] = (number >> 24) & 0xFF;
  bytes[offset + 5] = (number >> 16) & 0xFF;
  bytes[offset + 6] = (number >> 8) & 0xFF;
  bytes[offset + 7] = number & 0xFF;
}

uint32_t autograph_read_index(const uint8_t *bytes) {
  return get_uint32(bytes, 0);
}

size_t autograph_read_size(const uint8_t *bytes) {
  return get_uint64(bytes, 0);
}
